### PR TITLE
Add SRV Record Type

### PR DESCRIPTION
**Pull request by tlimoncelli**
Fixing https://github.com/StackExchange/dnscontrol/issues/56

### CODE REVIEW

**Comment by tlimoncelli** [comment]
I could use help fixing the diff mechanism and adding SRV records to the intergrationTests.

File: models/dns.go [hunk 1]
```diff
@@ -65,8 +65,10 @@ type RecordConfig struct {
 	Target   string            `json:"target"` // If a name, must end with "."
 	TTL      uint32            `json:"ttl,omitempty"`
 	Metadata map[string]string `json:"meta,omitempty"`
-	NameFQDN string            `json:"-"` // Must end with ".$origin". See below.
-	Priority uint16            `json:"priority,omitempty"`
+	NameFQDN string            `json:"-"`                  // Must end with ".$origin". See below.
+	Priority uint16            `json:"priority,omitempty"` // Used by SRV
```
Comment (captncraig) [inline models/dns.go]: Comment on `Priority` should probably be "used by SRV and MX"

File: pkg/js/helpers.js [hunk 2]
```diff
@@ -304,6 +313,33 @@ function addRecord(d,type,name,target,mods) {
     return rec;
 }
 
+function addRecordSRV(d,type,name,priority,weight,port,target,mods) {
+    var rec = {type: type, name: name, priority: priority, weight: weight, port: port, target: target, ttl:d.defaultTTL, meta:{}};
+    // for each modifier, decide based on type:
+    // - Function: call is with the record as the argument
+    // - Object: merge it into the metadata
+    // FIXME(tlim): Factor this code out to its own function.
```
Comment (captncraig) [inline pkg/js/helpers.js]: instead of having a separate `addRecordSRV` function, just do `var rec = addRecord(d, "SRV", name, target, mods)` then set weight / port and those fields on the returned record.

File: pkg/normalize/validate.go [hunk 3]
```diff
@@ -143,6 +144,8 @@ func checkTargets(rec *models.RecordConfig, domain string) (errs []error) {
 		}
 	case "ALIAS":
 		check(checkTarget(target))
+	case "SRV":
+		check(checkTarget(target))
```
Comment (captncraig) [inline pkg/normalize/validate.go]: Is a target of an IP address valid for SRV records? 

If so, I'm not sure I like how this validates. 

`SRV("@", 5, 5, 4242, "2.3.4.5")` is not valid but `SRV("@", 5, 5, 4242, "2.3.4.5.")` would be. But I don't know the rules. 

Should maybe the `checkTarget` function do more to make sure you are *not* pointing CNAME/MX/ALIAS/SRV at raw IP addresses?

File: providers/activedir/activedirProvider.go [hunk 4]
```diff
@@ -21,7 +21,7 @@ type adProvider struct {
 // Register with the dnscontrol system.
 //   This establishes the name (all caps), and the function to call to initialize it.
 func init() {
-	providers.RegisterDomainServiceProviderType("ACTIVEDIRECTORY_PS", newDNS)
+	providers.RegisterDomainServiceProviderType("ACTIVEDIRECTORY_PS", newDNS, providers.CanUseSRV)
```
Comment (captncraig) [inline providers/activedir/activedirProvider.go]: I am skeptical any of these providers are going to work as-is. They will need to read the weight/priority and other fields from the providers and marshal it into our struct format. And vice-versa. Maybe that bit is still WIP?

Comment (captncraig) [inline providers/activedir/activedirProvider.go]: We can and should add a conditional integration test as we add providers to the "support srv" club.

File: models/dns.go [hunk 5]
```diff
@@ -60,28 +61,71 @@ type DNSProviderConfig struct {
...
+	Priority    uint16            `json:"priority,omitempty"` // MX
+	SrvPriority uint16            `json:"SrvPriority,omitempty"`
+	SrvWeight   uint16            `json:"SrvWeight,omitempty"`
+	SrvPort     uint16            `json:"SrvPort,omitempty"`
 
 	Original interface{} `json:"-"` // Store pointer to provider-specific record object. Used in diffing.
 }
 
 func (r *RecordConfig) String() string {
 	content := fmt.Sprintf("%s %s %s %d", r.Type, r.NameFQDN, r.Target, r.TTL)
```
Comment (tlimoncelli) [inline models/dns.go]: Move this into the case statement. Eliminate the concats.

File: models/dns.go [hunk 6]
```diff
@@ -109,6 +153,13 @@ func (r *RecordConfig) ToRR() dns.RR {
...
+	case dns.TypeSRV:
+		// Has a Priority, Weight, and Port fields.
+		return &dns.SRV{Hdr: hdr,
+			Priority: r.SrvPriority,
+			Weight:   r.SrvWeight,
+			Port:     r.SrvPort,
+			Target:   r.Target}
 	case dns.TypeTXT:
 		// Assure no problems due to quoting/unquoting:
 		return &dns.TXT{Hdr: hdr, Txt: []string{r.Target}}
```
Comment (tlimoncelli) [inline models/dns.go]: the DefaultTTL processing below should be moved before the switches.

**Comment by tlimoncelli** [comment]
SRV records now work for these providers: BIND ROUTE53 GCLOUD NAMEDOTCOM

**Comment by tlimoncelli** [comment]
This is ready for review.

File: pkg/js/helpers.js [hunk 7]
```diff
@@ -312,6 +321,33 @@ function addRecord(d,type,name,target,mods) {
     return rec;
 }
 
+function addRecordSRV(d,type,name,srvpriority,srvweight,srvport,target,mods) {
+    var rec = {type: type, name: name, srvpriority: srvpriority, srvweight: srvweight, srvport: srvport, target: target, ttl:d.defaultTTL, meta:{}};
+    // for each modifier, decide based on type:
+    // - Function: call is with the record as the argument
+    // - Object: merge it into the metadata
+    // FIXME(tlim): Factor this code out to its own function.
```
Comment (captncraig) [inline pkg/js/helpers.js]: This does need to be factored into a common method and combined.

File: providers/dnsimple/dnsimpleProvider.go [hunk 8]
```diff
@@ -315,7 +315,7 @@ func newProvider(m map[string]string, metadata json.RawMessage) (*DnsimpleApi, e
 
 func init() {
 	providers.RegisterRegistrarType("DNSIMPLE", newReg)
-	providers.RegisterDomainServiceProviderType("DNSIMPLE", newDsp)
+	providers.RegisterDomainServiceProviderType("DNSIMPLE", newDsp, providers.CanUsePTR)
```
Comment (captncraig) [inline providers/dnsimple/dnsimpleProvider.go]: Intended to enable ptr here? This part of your merge troubles, or a typo?

File: models/dns.go [hunk 9]
```diff
@@ -60,74 +61,134 @@ type DNSProviderConfig struct {
...
+	}
 
-	// NB(tlim): An alternative way to do this would be
-	// to create the rr via: rr := TypeToRR[x]()
-	// then set the parameters. A benchmark may find that
-	// faster. This was faster to implement.
+	// We cheat by converting to a dns.RR and use the String() function.
+	// Sadly that function always includes a header, which we must strip out.
+	// TODO(tlim): Request the dns project add a function that returns
+	// the string without the header.
```
Comment (captncraig) [inline models/dns.go]: I'm not sure I love relying on the dns package for this, especially with the number of caveats we add. Maybe just implement our own switch?

File: models/dns.go [hunk 10]
```diff
@@ -60,74 +61,134 @@ type DNSProviderConfig struct {
...
-	Metadata map[string]string `json:"meta,omitempty"`
-	NameFQDN string            `json:"-"` // Must end with ".$origin". See below.
-	Priority uint16            `json:"priority,omitempty"`
+	Type        string            `json:"type"`
+	Name        string            `json:"name"`   // The short name. See below.
+	Target      string            `json:"target"` // If a name, must end with "."
+	TTL         uint32            `json:"ttl,omitempty"`
+	Metadata    map[string]string `json:"meta,omitempty"`
+	NameFQDN    string            `json:"-"`                  // Must end with ".$origin". See below.
+	Priority    uint16            `json:"priority,omitempty"` // MX
```
Comment (tlimoncelli) [inline models/dns.go]: I've added a FIXME().  That should be a different PR.

File: pkg/js/helpers.js (hunk 7 above)
Comment (tlimoncelli) [inline pkg/js/helpers.js]: I'm not confident in my javascript abilities to do that.  I'll file an issue and we'll get to this later.

File: models/dns.go (hunk 9 above)
Comment (tlimoncelli) [inline models/dns.go]: I'm ok with this.

File: models/dns.go (hunk 10 above)
Comment (tlimoncelli) [inline models/dns.go]: Fixed

File: providers/namedotcom/namedotcomProvider.go [hunk 11]
```diff
@@ -41,7 +41,8 @@ func newProvider(conf map[string]string) (*nameDotCom, error) {
 
 func init() {
 	providers.RegisterRegistrarType("NAMEDOTCOM", newReg)
-	providers.RegisterDomainServiceProviderType("NAMEDOTCOM", newDsp, providers.CanUseAlias)
+	providers.RegisterDomainServiceProviderType("NAMEDOTCOM", newDsp, providers.CanUseSRV)
```
Comment (whs) [inline providers/namedotcom/namedotcomProvider.go]: Seems that providers.CanUseAlias was ommitted?

Comment (tlimoncelli) [inline providers/namedotcom/namedotcomProvider.go]: Oops, that was a merge error. I've added back CanUseAlias.
